### PR TITLE
API for Access Token Validation in OAuth/OIDC

### PR DESCRIPTION
**Pull request by kdorosh**
# Description

API for Gloo to handle other parts of the [OAuth2.0](https://tools.ietf.org/html/rfc6749), [OIDC](https://openid.net/specs/openid-connect-core-1_0.html), [token introspection](https://tools.ietf.org/html/rfc7662), and [bearer token](https://tools.ietf.org/rfc/rfc6750.txt) specs in our external auth server implementation. Currently we support an OIDC authorization code flow, but users have requested that Gloo handle other parts of the OAuth2.0 spec as well. In this iteration, we are adding access token validation via token introspection and making the userinfo response available to users in extauth go plugins.

# Context

This PR deprecates the old OAuth config as it really only supported OIDC authorization code flows and left little room to cleanly implement other parts of the specs (both OAuth2.0 and OIDC).

The new API allows us to easily add more parts of the spec incrementally (such as `OidcImplicit` and `OidcHybrid` flows, as well as other ways to validate the access token, such as `Jwt` or `Hmac`) without making awkward additions to the previous message that would require setting one field completely obsolesce large portions of the other fields.

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make update-deps generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo [contribution guide](https://docs.solo.io/gloo/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 1]
```diff
@@ -193,7 +194,51 @@ message BasicAuth {
...
+
+  // a callback path relative to app url that will be used for OIDC callbacks.
+  // needs to not be used by the application
+  string callback_path = 5 [deprecated=true];
+
+  // Scopes to request in addition to openid scope.
+  repeated string scopes = 6 [deprecated=true];
+}
+
+message OAuth2 {
```
Comment (kdorosh) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:224]: TODO: make sure API accounts for machine to machine authn

**Comment by keiransteele** [comment]
Hi All, very interested in this so that we can do token introspection.

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto (hunk 1 above)
Comment (kdorosh) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:224]: we can just put client credentials into the new oneof and will be set there 👍

**Comment by kdorosh** [comment]
/kick virtual service regression test flake, https://console.cloud.google.com/cloud-build/builds/283cdfd1-6a10-4d8e-8971-d1db978549f2?project=solo-public

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 2]
```diff
@@ -205,18 +250,45 @@ message OAuth {
...
+  repeated string scopes = 7;
+
+  // TODO(kdorosh) add optional configuration for validation of the access token received during the OIDC flow
+  // AccessTokenValidation access_token_validation = 8;
+}
+
+message AccessTokenValidation {
+
+  // fields 2 and 3 reserved for jwt and hmac validation
+  reserved 2, 3;
```
Comment (kdorosh) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:273]: let's keep reserved for things that were previously deprecated

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 3]
```diff
@@ -205,18 +250,45 @@ message OAuth {
...
+  // the url for the OIDC userinfo endpoint.
+  // if provided, the (opaque) access token provided or received from the oauth endpoint
+  // will be queried and the userinfo response (or cached response) will be put in the `AuthorizationRequest` state.
+  // this can be useful to leverage the userinfo response in, for example, an extauth server plugin.
+  string userinfo_url = 4;
+
+  // how long the token introspection and/or userinfo endpoint response for a specific access token should be kept
+  // in the in-memory cache. the result will be invalidated at this timeout, or at "exp" time from the
+  // introspection result, whichever comes sooner. if omitted, defaults to 10 minutes.
+  google.protobuf.Duration cache_timeout = 5 [(gogoproto.stdduration) = true];
```
Comment (kdorosh) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:291]: support zero value for never evict

**Comment by kdorosh** [comment]
/kick mix of valid and invalid vs flake https://console.cloud.google.com/cloud-build/builds/05004826-8c57-482b-9ae7-544a470bab9c?project=907026730415

**Comment by kdorosh** [comment]
/kick mix of valid and invalid vs flake https://console.cloud.google.com/cloud-build/builds/54f72dae-7ff0-4fa5-8c7c-cba166e4dac3?project=solo-public

**Comment by kdorosh** [comment]
/kick e2e test fail https://storage.googleapis.com/solo-public-build-logs/logs.html?buildid=df006605-3360-492c-ac2d-d15951291c8d

**Pull request merged**